### PR TITLE
Upgrade to jsonld ^5.0.2.

### PR DESCRIPTION
**Pull request by aljones15**
Addresses: https://github.com/digitalbazaar/bedrock-ledger-node/issues/101

- [x] Look for deprecated or changed parameters with `jsonld.cannonize`
  - [x] algorithm 'URDNA2015'
  - [x] documentLoader
  - [x] format 'application/n-quads'
- [x] ensure test project actually hits the code paths that use `jsonld`
- [x] investigate CI node v14 failure

PR Concerns:

`jsonld: ^2.0.0` cannonize:

```js
 {
    "@context": "https://w3id.org/webledger/v1",
    "type": "WebLedgerConfigurationEvent",
    "ledgerConfiguration": {
      "@context": [
        "https://w3id.org/webledger/v1",
        "https://w3id.org/zcap/v1",
        "https://w3id.org/security/suites/ed25519-2020/v1"
      ],   
      "type": "WebLedgerConfiguration",
      "ledger": "did:v1:uuid:eb8c22dc-bde6-4315-92e2-59bd3f3c7d59",
      "consensusMethod": "UnilateralConsensus2017",
      "sequence": 0,
      "proof": {
        "type": "Ed25519Signature2020",
        "created": "2021-10-22T16:06:41Z",
        "verificationMethod": "https://good.example/keys/authorized-key-1",
        "proofPurpose": "capabilityInvocation",
        "capability": "did:v1:uuid:eb8c22dc-bde6-4315-92e2-59bd3f3c7d59",
        "invocationTarget": "did:v1:uuid:eb8c22dc-bde6-4315-92e2-59bd3f3c7d59",
        "capabilityAction": "write",
        "proofValue": "z2UUEMNjkhVzpMKoqzbgeNYoSi2oe6fii7dT3ZQRBho64JGFnt6gbm9YtD8PG39wQPbdy88btMBre2hnXCctGm9Xq"
      }    
    }    
  },
  "canonized": "_:c14n1 <http://purl.org/dc/terms/created> \"2021-10-22T16:06:41Z\"^^<http://www.w3.org/2001/XMLSchema#dateTime> _:c14n0 .\n_:c14n1
<http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://w3id.org/security#Ed25519Signature2020> _:c14n0 .\n_:c14n1 <https://w3id.org/                
security#capability> <did:v1:uuid:eb8c22dc-bde6-4315-92e2-59bd3f3c7d59> _:c14n0 .\n_:c14n1 <https://w3id.org/security#capabilityAction> \"write\" _:  
c14n0 .\n_:c14n1 <https://w3id.org/security#invocationTarget> <did:v1:uuid:eb8c22dc-bde6-4315-92e2-59bd3f3c7d59> _:c14n0 .\n_:c14n1 <https://w3id.org/
security#proofPurpose> <https://w3id.org/security#capabilityInvocationMethod> _:c14n0 .\n_:c14n1 <https://w3id.org/security#proofValue>              
\"z2UUEMNjkhVzpMKoqzbgeNYoSi2oe6fii7dT3ZQRBho64JGFnt6gbm9YtD8PG39wQPbdy88btMBre2hnXCctGm9Xq\"^^<https://w3id.org/security#multibase> _:c14n0 .\n_:c14n1
<https://w3id.org/security#verificationMethod> <https://good.example/keys/authorized-key-1> _:c14n0 .\n_:c14n2 <http://www.w3.org/1999/02/22-rdf-syntax-
ns#type> <https://w3id.org/webledger#WebLedgerConfiguration> _:c14n4 .\n_:c14n2 <https://w3id.org/json-ld-patch#sequence> \"0\"^^<http://www.w3.org/
2001/XMLSchema#integer> _:c14n4 .\n_:c14n2 <https://w3id.org/security#proof> _:c14n0 _:c14n4 .\n_:c14n2 <https://w3id.org/webledger#ledger> <did:v1: 
uuid:eb8c22dc-bde6-4315-92e2-59bd3f3c7d59> _:c14n4 .\n_:c14n3 <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://w3id.org/                
webledger#WebLedgerConfigurationEvent> .\n_:c14n3 <https://w3id.org/webledger#ledgerConfiguration> _:c14n4 .\n"
}
```

`jsonld: ^5.0.1`

```js
    "@context": "https://w3id.org/webledger/v1",
    "type": "WebLedgerConfigurationEvent",
    "ledgerConfiguration": {
      "@context": [
        "https://w3id.org/webledger/v1",
        "https://w3id.org/zcap/v1",
        "https://w3id.org/security/suites/ed25519-2020/v1"
      ],   
      "type": "WebLedgerConfiguration",
      "ledger": "did:v1:uuid:eb8c22dc-bde6-4315-92e2-59bd3f3c7d59",
      "consensusMethod": "UnilateralConsensus2017",
      "sequence": 0,
      "proof": {
        "type": "Ed25519Signature2020",
        "created": "2021-10-22T16:03:31Z",
        "verificationMethod": "https://good.example/keys/authorized-key-1",
        "proofPurpose": "capabilityInvocation",
        "capability": "did:v1:uuid:eb8c22dc-bde6-4315-92e2-59bd3f3c7d59",
        "invocationTarget": "did:v1:uuid:eb8c22dc-bde6-4315-92e2-59bd3f3c7d59",
        "capabilityAction": "write",
        "proofValue": "z31PNNwqwRnr9XGt8xAhHgyLSaSkzoYumXmGsFRmbyn3MHnkH36CovZAFbxhQfJ2ZeRUj8tyjmGfGUpkiSYjYCzNS"
      }    
    }    
  }
  "canonized": "_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://w3id.org/webledger#WebLedgerConfiguration> _:c14n2 .\n_:c14n0        
<https://w3id.org/json-ld-patch#sequence> \"0\"^^<http://www.w3.org/2001/XMLSchema#integer> _:c14n2 .\n_:c14n0 <https://w3id.org/security#proof> _:     c14n3 _:c14n2 .\n_:c14n0 <https://w3id.org/webledger#ledger> <did:v1:uuid:eb8c22dc-bde6-4315-92e2-59bd3f3c7d59> _:c14n2 .\n_:c14n1 <http://www.w3.org/  1999/02/22-rdf-syntax-ns#type> <https://w3id.org/webledger#WebLedgerConfigurationEvent> .\n_:c14n1 <https://w3id.org/webledger#ledgerConfiguration> _:
c14n2 .\n_:c14n4 <http://purl.org/dc/terms/created> \"2021-10-22T16:03:31Z\"^^<http://www.w3.org/2001/XMLSchema#dateTime> _:c14n3 .\n_:c14n4 <http://
www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://w3id.org/security#Ed25519Signature2020> _:c14n3 .\n_:c14n4 <https://w3id.org/security#capability>
<did:v1:uuid:eb8c22dc-bde6-4315-92e2-59bd3f3c7d59> _:c14n3 .\n_:c14n4 <https://w3id.org/security#capabilityAction> \"write\" _:c14n3 .\n_:c14n4
<https://w3id.org/security#invocationTarget> <did:v1:uuid:eb8c22dc-bde6-4315-92e2-59bd3f3c7d59> _:c14n3 .\n_:c14n4 <https://w3id.org/
security#proofPurpose> <https://w3id.org/security#capabilityInvocationMethod> _:c14n3 .\n_:c14n4 <https://w3id.org/security#proofValue>
\"z31PNNwqwRnr9XGt8xAhHgyLSaSkzoYumXmGsFRmbyn3MHnkH36CovZAFbxhQfJ2ZeRUj8tyjmGfGUpkiSYjYCzNS\"^^<https://w3id.org/security#multibase> _:c14n3 .\n_:c14n4
<https://w3id.org/security#verificationMethod> <https://good.example/keys/authorized-key-1> _:c14n3 .\n"
```

`WebLedgerConfigurationEvent` cannonizations are not the same.
1. Is the `jsonld: ^5.0.0` cannonize in `n-quads` format?
2. Have the documentLoader or other docs changed?
3. We don't test for cannonizalization.

Code path affected by this upgrade:

https://github.com/digitalbazaar/bedrock-ledger-node/blob/4c4f0be9a28f5315eb30c50c147b3756cc30ba0b/lib/rdfCanonizeAndHash.js#L16-L27

### CODE REVIEW

**Review by dlongley — APPROVED** [review]
Thanks!

**Comment by davidlehn** [comment]
I checked the canonize issue and that's a false alarm as far as I can tell.  The event examples above were signed at different times resulting in different `created` and `proofValue` fields.  The canonize algorithm therefore produces different outputs.  The version of `jsonld` is unrelated to that difference.

Just to double check, I ran both inputs through expansion and canonization for both `jsonld` versions.  The outputs all match as expected.

(I think there are JSON-LD edge cases that would have issues between `jsonld` versions, but I don't think ledger code and data hits those issues.)

**Comment by gannan08** [comment]
> I checked the canonize issue and that's a false alarm as far as I can tell. The event examples above were signed at different times resulting in different `created` and `proofValue` fields. The canonize algorithm therefore produces different outputs. The version of `jsonld` is unrelated to that difference.
> 
> Just to double check, I ran both inputs through expansion and canonization for both `jsonld` versions. The outputs all match as expected.
> 
> (I think there are JSON-LD edge cases that would have issues between `jsonld` versions, but I don't think ledger code and data hits those issues.)

Thank you very much for looking into this.

**Review by gannan08 — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,10 @@
 # bedrock-ledger-node ChangeLog
 
+## 15.2.0 - TBD
+
+### Changed
+- Upgraded `jsonld` to `^5.2.0`.
```
[the remark this answers is not in the document]
Comment (aljones15) [reply CHANGELOG.md:6]: @davidlehn @dlongley are we sure this is a major change? @davidlehn seems to be suggesting the canonize outputs are similar between versions and the differences I saw is mostly do to different values in the created timestamp.

File: .gitignore [hunk 2]
```diff
@@ -11,3 +12,4 @@ coverage
 node_modules
 npm-debug.log
 reports
+package-lock.json
```
[the remark this answers is not in the document]
Comment (aljones15) [reply .gitignore]: Removed `package-lock.json` from `.gitignore` here: https://github.com/digitalbazaar/bedrock-ledger-node/pull/102/commits/2103cc60f7576513bd675edcf5b4f729a25b986e

**Review by mattcollier — APPROVED** [review]
Thanks to all that worked on this.

**Review by davidlehn — APPROVED** [review]
Approving.  I don't know how cautious to be with versioning to help deal with any potential issues due to expanding/canonizing data.  Any problems could be quite difficult to deal with so it would be nice to have tests to detect potential issues.  But it might be a lot of effort to figure out the edge cases that could have problems now or in the future.

File: CHANGELOG.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (davidlehn) [reply CHANGELOG.md:6]: @dlongley @gannan08 @mattcollier: It appears this could be just a minor update based on the tests (including those added in https://github.com/digitalbazaar/bedrock-ledger-node/pull/105 / https://github.com/digitalbazaar/bedrock-ledger-node/pull/106). Are there other considerations that should make this a major release?  For instance, just to avoid potential unknown issues until a manual version update occurs.

[the remark this answers is not in the document]
Comment (davidlehn) [reply CHANGELOG.md:6]: @aljones15 In a call it was decided to use a major version. Basically just adds a bit of safety since we don't know how to test for potential unknown issues.

Comment (davidlehn) [inline CHANGELOG.md:6]:
```suggestion
- **BREAKING**: Upgraded `jsonld` to `^5.2.0`. Marked as a breaking major change to help avoid any unknown upgrading issues in live systems due to potential JSON-LD edge case processing changes. No issues are expected in normal use cases.
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,10 @@
 # bedrock-ledger-node ChangeLog
 
+## 15.2.0 - TBD
```
Comment (davidlehn) [inline CHANGELOG.md:3]:
```suggestion
## 16.0.0 - TBD
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,11 @@
 # bedrock-ledger-node ChangeLog
 
+## 16.0.0 - TBD
+
+### Changed
+- **BREAKING**: Upgraded `jsonld` to `^5.2.0` this might cause the `canonizedBytes`
+  & `hash` of records to change.
```
Comment (davidlehn) [inline CHANGELOG.md:7]: Could use a reword as is, but also I think readers won't know what to do with that statement.  Perhaps a hybrid with the text I had proposed could help.  I tried to not delve into the details and suggest we don't think there are issues.
```suggestion
- **BREAKING**: Upgraded `jsonld` to `^5.2.0`. Marked as a breaking major change to help avoid any unknown upgrading issues in live systems due to potential JSON-LD edge case processing changes. No issues are expected in normal use cases. If such an issue exists, it might cause the `canonizedBytes` and `hash` of records to change.
```

**Comment by aljones15** [comment]
please note: do not delete the branch `upgrade-jsonld` right now. Some projects are depending on it.
